Add Nginx Bad Bot Blocker & update config

tychuang1211 opened the pull request:
This PR introduces:

- Replacing the default nginx config with demo configuration
- Integrating [nginx-ultimate-bad-bot-blocker](https://github.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker)

For bad-bot blocker updates, I'd suggest running a cron job on the host machine, since:
- Running cron (as a daemon) within Docker is discouraged, because Docker only monitors the main PID
- The [jonasal/nginx-certbot](https://github.com/JonasAlfredsson/docker-nginx-certbot) image already manages multiple foreground processes (Nginx and certbot) through a parent process that monitors both
- The blocker update requires a Nginx reload. While this can be triggered from a separate container using shared control mechanisms, it's generally not considered good design due to the added complexity and coupling between containers

tychuang1211 committed:
Add Nginx Bad Bot Blocker & update config

tychuang1211 commented on nginx/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+FROM jonasal/nginx-certbot:5.4
```
> ℹ️ Please use a [specific tag](https://github.com/JonasAlfredsson/docker-nginx-certbot/blob/master/docs/dockerhub_tags.md) when doing a Docker pull, since :latest might not always be 100% stable.

from [docker-nginx-certbot readme](https://github.com/JonasAlfredsson/docker-nginx-certbot?tab=readme-ov-file#docker-nginx-certbot)

tychuang1211 commented on nginx/user_conf.d/default.conf:
```diff
@@ -19,21 +43,87 @@ server {
     # ssl_session_cache dfine in stream and http
     ssl_session_tickets off;
+    ssl_session_cache shared:SSL:30m;
+    # this is the internal Docker DNS, cache only for 30s
+    resolver 127.0.0.11 valid=30s;
+    ssl_stapling on;
+    ssl_stapling_verify on;
+    ssl_buffer_size 8k;
+
+    # Let nginx start if upstream host is unavailable or down
+    set $ckan ckan:5000;
```
Allow Nginx to start even if the upstream (ckan) service is unavailable.
Without this, Nginx may fail to start/restart when the upstream host is down.
Enabling dynamic resolving ensures Nginx can still start independently.

u10313335 commented on nginx/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+RUN apt-get update &&  apt-get install --no-install-recommends --no-install-suggests -y \
+ wget
+
+# https://github.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/blob/master/MANUAL-CONFIGURATION.md
```
Is there any reason for doing manual installation rather than using [auto installation scripts](https://github.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/blob/master/AUTO-CONFIGURATION.md) here?

tychuang1211 replied:
I've checked the auto installation scripts. `install-ngxblocker` works for installing the necessary files, and can replace lines 8-19 if needed. However I can't get `setup-ngxblocker` to work properly, so the include directives still need to be added manually ([as shown here](https://github.com/depositar/depositar-docker/pull/1/files/f8f2fdff818e39bdc14e36c33d7987f584ae8815#diff-c2e61af6cd6e0bf008dadceb0953247dc3203f8a6e298dc89d16e273ef6a9c24R60-R64)).

u10313335 replied:
> However I can't get `setup-ngxblocker` to work properly, so the include directives still need to be added manually ([as shown here](https://github.com/depositar/depositar-docker/pull/1/files/f8f2fdff818e39bdc14e36c33d7987f584ae8815#diff-c2e61af6cd6e0bf008dadceb0953247dc3203f8a6e298dc89d16e273ef6a9c24R60-R64)).

Could you try to specify the configuration directory via `./setup-ngxblocker -c /etc/nginx/user_conf.d`?

tychuang1211 replied:
> > However I can't get `setup-ngxblocker` to work properly, so the include directives still need to be added manually ([as shown here](https://github.com/depositar/depositar-docker/pull/1/files/f8f2fdff818e39bdc14e36c33d7987f584ae8815#diff-c2e61af6cd6e0bf008dadceb0953247dc3203f8a6e298dc89d16e273ef6a9c24R60-R64)).
> 
> Could you try to specify the configuration directory via `./setup-ngxblocker -c /etc/nginx/user_conf.d`?

I tried that, but it still complains about missing vhost files, even when I add the `-z` flag (`[ -z ] : NO configuration of vhosts (default: configure files in /etc/nginx/sites-available)`).
```
no vhost files in: [ /etc/nginx/sites-available/*.conf ] => exiting.
```
I also tried `./setup-ngxblocker -v /etc/nginx/user_conf.d -e conf`, but it didn’t help.
I also found a similar issue reported [here](https://github.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/issues/501).

u10313335 replied:
Please rename the `default.conf` in the `nginx/user_conf.d` to `default.vhost`, then run the command `./setup-ngxblocker -v /etc/nginx/conf.d`.

PS. we rename the `default.conf` because there are other `.conf` in the `/etc/nginx/conf.d`.

u10313335 replied:
This is because the `default.conf` only appears in the `/etc/nginx/conf.d` (the real path used by NGINX) when the container is running (via a [symbolic link](https://github.com/JonasAlfredsson/docker-nginx-certbot/blob/master/src/scripts/util.sh#L254)). Therefore, the `setup-ngxblocker` doesn't work during image creation.

So let's keep the lines added by the `setup-ngxblocker` script in the `default.conf`, just like what you've written so far.

u10313335 commented:
> For bad-bot blocker updates, I'd suggest running a cron job on the host machine, since:
> 
> * Running cron (as a daemon) within Docker is discouraged, because Docker only monitors the main PID
> * The [jonasal/nginx-certbot](https://github.com/JonasAlfredsson/docker-nginx-certbot) image already manages multiple foreground processes (Nginx and certbot) through a parent process that monitors both
> * The blocker update requires a Nginx reload. While this can be triggered from a separate container using shared control mechanisms, it's generally not considered good design due to the added complexity and coupling between containers

My concern is that the scheduling part will not be dockerized if we run a cron job on the host machine, which hurts the benefits of Docker (easy deployment and high probability). Instead, I'm wondering whether something like the [ofelia](https://github.com/mcuadros/ofelia) can help us schedule tasks inside Docker containers.

tychuang1211 commented:
> > For bad-bot blocker updates, I'd suggest running a cron job on the host machine, since:
> > 
> > * Running cron (as a daemon) within Docker is discouraged, because Docker only monitors the main PID
> > * The [jonasal/nginx-certbot](https://github.com/JonasAlfredsson/docker-nginx-certbot) image already manages multiple foreground processes (Nginx and certbot) through a parent process that monitors both
> > * The blocker update requires a Nginx reload. While this can be triggered from a separate container using shared control mechanisms, it's generally not considered good design due to the added complexity and coupling between containers
> 
> My concern is that the scheduling part will not be dockerized if we run a cron job on the host machine, which hurts the benefits of Docker (easy deployment and high probability). Instead, I'm wondering whether something like the [ofelia](https://github.com/mcuadros/ofelia) can help us schedule tasks inside Docker containers.

Sounds good, I’ll take a look at ofelia and see if it could help.

u10313335 commented on nginx/user_conf.d/default.conf:
```diff
@@ -19,21 +43,87 @@ server {
+        }
+
+        proxy_pass http://$ckan$request_uri;
+        include /etc/nginx/user_conf.d/ckan/depositar_params;
```
```suggestion
        proxy_set_header X-Forwarded-For $remote_addr;
        proxy_set_header Host $host;
        proxy_read_timeout 600s;
        # when a client closes the connection then keep the channel to uwsgi open. Otherwise uwsgi throws an IOError
        uwsgi_ignore_client_abort on;
```

u10313335 commented on nginx/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+ENV NGINX_DIR=/etc/nginx
+
+RUN apt-get update &&  apt-get install --no-install-recommends --no-install-suggests -y \
+ wget
```
```suggestion
RUN apt-get update && apt-get install --no-install-recommends --no-install-suggests -y \
 wget curl dnsutils
```

The `curl` and `dig` (dnsutils) are needed to run the auto installation script.

u10313335 commented on nginx/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+RUN wget https://raw.githubusercontent.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/master/bots.d/custom-bad-referrers.conf -O ${NGINX_DIR}/bots.d/custom-bad-referrers.conf
+RUN wget https://raw.githubusercontent.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/master/bots.d/blacklist-ips.conf -O ${NGINX_DIR}/bots.d/blacklist-ips.conf
+RUN wget https://raw.githubusercontent.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/master/bots.d/bad-referrer-words.conf -O ${NGINX_DIR}/bots.d/bad-referrer-words.conf
+RUN wget https://raw.githubusercontent.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/master/conf.d/botblocker-nginx-settings.conf -O ${NGINX_DIR}/conf.d/botblocker-nginx-settings.conf
```
Run the auto installation script:

```suggestion
# https://github.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/blob/master/AUTO-CONFIGURATION.md
RUN wget https://raw.githubusercontent.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/master/install-ngxblocker -O /usr/local/sbin/install-ngxblocker
WORKDIR /usr/local/sbin
RUN chmod +x install-ngxblocker
RUN install-ngxblocker -x
```

u10313335 commented on nginx/user_conf.d/ckan/depositar_params:
Move this file to `nginx/user_conf.d/depositar_params` to reference it by `conf.d/depositar_params` in the nginx config.

tychuang1211 committed:
Use auto installation script for nginx-bad-bot-blocker

Co-authored-by: Sol Lee <u103133.u103135@gmail.com>

tychuang1211 committed:
Move `depositar_params` to `user_conf.d`

tychuang1211 commented:
Thanks for the suggestions. I've applied them in the last two commits:

1. Switched to using auto installation script for nginx-bad-bot-blocker.
 2. Moved `nginx/user_conf.d/ckan/depositar_params` to `nginx/user_conf.d/depositar_params`.

Let me know if there’s anything else I should adjust.

tychuang1211 commented:
> > > For bad-bot blocker updates, I'd suggest running a cron job on the host machine, since:
> > > 
> > > * Running cron (as a daemon) within Docker is discouraged, because Docker only monitors the main PID
> > > * The [jonasal/nginx-certbot](https://github.com/JonasAlfredsson/docker-nginx-certbot) image already manages multiple foreground processes (Nginx and certbot) through a parent process that monitors both
> > > * The blocker update requires a Nginx reload. While this can be triggered from a separate container using shared control mechanisms, it's generally not considered good design due to the added complexity and coupling between containers
> > 
> > 
> > My concern is that the scheduling part will not be dockerized if we run a cron job on the host machine, which hurts the benefits of Docker (easy deployment and high probability). Instead, I'm wondering whether something like the [ofelia](https://github.com/mcuadros/ofelia) can help us schedule tasks inside Docker containers.
> 
> Sounds good, I’ll take a look at ofelia and see if it could help.

@u10313335 I looked into ofelia, it seems like it could work for our use case.
Do you think I should commit the changes directly to this PR, or would it be better to open a separate one?

u10313335 commented:
> > > > For bad-bot blocker updates, I'd suggest running a cron job on the host machine, since:
> > > > 
> > > > * Running cron (as a daemon) within Docker is discouraged, because Docker only monitors the main PID
> > > > * The [jonasal/nginx-certbot](https://github.com/JonasAlfredsson/docker-nginx-certbot) image already manages multiple foreground processes (Nginx and certbot) through a parent process that monitors both
> > > > * The blocker update requires a Nginx reload. While this can be triggered from a separate container using shared control mechanisms, it's generally not considered good design due to the added complexity and coupling between containers
> > > 
> > > 
> > > My concern is that the scheduling part will not be dockerized if we run a cron job on the host machine, which hurts the benefits of Docker (easy deployment and high probability). Instead, I'm wondering whether something like the [ofelia](https://github.com/mcuadros/ofelia) can help us schedule tasks inside Docker containers.
> > 
> > 
> > Sounds good, I’ll take a look at ofelia and see if it could help.
> 
> @u10313335 I looked into ofelia, it seems like it could work for our use case. Do you think I should commit the changes directly to this PR, or would it be better to open a separate one?

Another PR is good to me. I will merge this one first.
